feat: the real implementation of deferred on the old codebase that's no longer valid and superseded by yet another PR built on top of React Router where the real deferred implementation actually lives (#3434)
remix-run/remix: docs/guides/streaming.md

---
title: Streaming
description: When, why, and how to stream with React 18 and Remix's deferred API.
---

# Streaming

Remix supports the [web streaming API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) as a first-class citizen. Additionally, JavaScript server runtimes have support for streaming responses to the client.

> One problem streaming in React 18 solves is for lazy-loading chunks of your code you don't want in your server render (for example, if a component doesn't support server-rendering). Once you enable React 18 streaming, you can use [`React.lazy`][react-lazy] on the server and the fallback will be rendered on the server and the client will fetch the code and render the components.

Author: @kentcdodds I don't believe this is 100% accurate. React.lazy *will* server render in React 18 when using one of the streaming renderer's instead of renderToString. We will be requiring one of those streaming renderers when using deferred, so I would re-word this or remove it.

Reviewer: Wait. To be clear... If you're using `renderToString` (even with React 18) you can't use `React.lazy` right? That's what I'm trying to communicate here. That enabling streaming by using `renderToPipeableStream` is required if you want to use `React.lazy` on the server render. Am I mistaken?

Author: It's correct that `renderToString` will bail out on the server due to usage of streaming features (suspense) without using a streaming compatible rendering method . That is a byproduct of attempting to use streaming features in an invalid context and isn't really a "streaming feature", more of a byproduct of improper use and isn't behavior I would encourage relying on in general as as soon as you do actually start streaming that use-case breaks.

Additionally, the API that Remix exposes for this is extremely ergonomic. You can literally switch between whether something is going to be derferred based on whether you include the `await` keyword:

```tsx
return deferred<LoaderData>({
  // not deferred:
  packageLocation: await packageLocationPromise,
  // deferred:
  packageLocation: packageLocationPromise,
});
```

Because of this, you can A/B test deferring, or even determine whether to defer based on the user or data being requested:

```tsx
export const loader: LoaderFunction = ({
  request,
  params,
}) => {
  const packageLocationPromise = getPackageLocation(
    params.packageId
  );
  const shouldDefer = await shouldDeferPackageLocation(
    request,
    params.packageId
  );

  return deferred<LoaderData>({
    packageLocation: shouldDefer
      ? packageLocationPromise
      : await packageLocationPromise,
  });
};
```

That `shouldDeferPackageLocation` could be implemented to check the user making the request, whether the package location data is in a cache, the status of an A/B test, or whatever else you want. This is pretty sweet 🍭

## When to not use streaming

> TODO: Talk about no streaming on action reloads.

Reviewer: @jacob-ebey, can you help me come up with an example of skipping deferring on revalidations?

Reviewer 2: Jacob should confirm but I believe `deferred` should be automatically "disabled" on revalidations as we just await all the promises before resolving the `loader` if it's not an initial request

          pipe(body);
        },
      }
    );
  });
}
```

<details>
  <summary>For a more complete example, expand this</summary>

This handles errors and properly disables streaming for bots which you typically want to force waiting so you can display all the content for SEO purposes.

```tsx filename=app/entry.server.tsx
import { PassThrough } from "stream";
import { renderToPipeableStream } from "react-dom/server";
import { RemixServer } from "@remix-run/react";
import { Response } from "@remix-run/node";
import type {
  EntryContext,
  Headers,
} from "@remix-run/node";
import isbot from "isbot";

const ABORT_DELAY = 5000;

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const callbackName = isbot(
    request.headers.get("user-agent")
  )
    ? "onAllReady"
    : "onShellReady";

  return new Promise((resolve, reject) => {
    let didError = false;

    const { pipe, abort } = renderToPipeableStream(
      <RemixServer
        context={remixContext}
        url={request.url}
      />,
      {
        [callbackName]() {
          let body = new PassThrough();

>           responseHeaders.set("Content-Type", "text/html");

Reviewer 3: Lighthouse audits say that you should add the `charset` in the first 1024 bytes streamed from the server. This can also be done via a meta tag but I feel like this is the easiest place to add it. 

```suggestion
          responseHeaders.set("Content-Type", "text/html; charset=UTF-8");
```

Reviewer: Perhaps we should update all the example to do this 🤔

Reviewer 4: According to @mathiasbynens, it's a best practice to use the `<meta charset="utf-8">` (see https://github.com/GoogleChrome/lighthouse/issues/10023#issuecomment-558521860)

Reviewer 3: Ok. I can see the argument to keep things as is. 

I using a CCS-in-JS library (I know it’s something the Remix community thinks should die) which automatically injects style tags into the top of the head in document. In my case, this bumped the meta tag to be too low in the document to work. Again, I understand that is an edge case so I will close this.

          resolve(
            new Response(body, {
              status: didError ? 500 : responseStatusCode,
              headers: responseHeaders,
            })
          );

>           pipe(body);

Reviewer 3: It's a best practice to add the `<!DOCTYPE html>` tag to prevent browser from going into "quirks mode". 

```suggestion
          body.write("<!DOCTYPE html>");
          pipe(body);
```

---
title: Streaming
description: When, why, and how to stream with React 18 and Remix's deferred API.
---

# Streaming

Remix supports the [web streaming API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) as a first-class citizen. Additionally, JavaScript server runtimes have support for streaming responses to the client.

One problem streaming in React 18 solves is for lazy-loading chunks of your code you don't want in your server render (for example, if a component doesn't support server-rendering). Once you enable React 18 streaming, you can use [`React.lazy`][react-lazy] on the server and the fallback will be rendered on the server and the client will fetch the code and render the components.

But there's a more nuanced problem that streaming solves. Let's look into that:

## The problem

Imagine a scenario where one of your routes' loaders needs to retrieve some data that for one reason or another is quite slow. For example, let's say you're showing the user the location of a package that's being delivered to their home:

```tsx

> import type { LoaderFunction } from "@remix-run/node";
> import { json } from "@remix-run/node";

Reviewer 4: ```suggestion
import type { LoaderFunction } from "@remix-run/node"; // or cloudflare/deno
import { json } from "@remix-run/node"; // or cloudflare/deno
```

            })
          );
          pipe(body);
        },
        onShellError(err: unknown) {
          reject(err);
        },
        onError(error: unknown) {
          didError = true;
          console.error(error);
        },
      }
    );
    setTimeout(abort, ABORT_DELAY);
  });
}
```

</details>

Then on the client you need to make sure you're hydrating properly with the React 18 `hydrateRoot` API:

```tsx filename=app/entry.client.tsx lines=[2,4]
import { RemixBrowser } from "@remix-run/react";
import { hydrateRoot } from "react-dom/client";

hydrateRoot(document, <RemixBrowser />);
```

With just that in place, you're unlikely to see any significant performance improvement. But with that alone you can now use [`React.lazy`][react-lazy] to SSR components but delay hydration on the client. This can open up network bandwidth for more critical things like styles, images, and fonts leading to a better LCP and TTI.

### Using `deferred`

With React streaming setup, now you can start adding `Deferred` usage for your slow data requests where you'd rather render a fallback UI. Let's do that for our example above:

```tsx lines=[3,5,6,11,18,23,33-46]

> import type { LoaderFunction } from "@remix-run/node";
> import { deferred } from "@remix-run/node";

Reviewer 4: ```suggestion
import type { LoaderFunction } from "@remix-run/node"; // or cloudflare/deno
import { deferred } from "@remix-run/node"; // or cloudflare/deno
```

        context={remixContext}
        url={request.url}
      />,
      {
        onShellReady() {
          const body = new PassThrough();

          responseHeaders.set("Content-Type", "text/html");

          resolve(
            new Response(body, {
              status: responseStatusCode,
              headers: responseHeaders,
            })
          );
          pipe(body);
        },
      }
    );
  });
}
```

<details>
  <summary>For a more complete example, expand this</summary>

This handles errors and properly disables streaming for bots which you typically want to force waiting so you can display all the content for SEO purposes.

```tsx filename=app/entry.server.tsx
import { PassThrough } from "stream";
import { renderToPipeableStream } from "react-dom/server";
import { RemixServer } from "@remix-run/react";
import { Response } from "@remix-run/node"; // or cloudflare/deno
import type {
  EntryContext,
  Headers,
} from "@remix-run/node"; // or cloudflare/deno
import isbot from "isbot";

const ABORT_DELAY = 5000;

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {

>   const callbackName = isbot(

Reviewer 5: Please branch the code example below. Something like:

```ts
if (isbot(...)) {
  return serveTheBots(req);
} else {
  return serveBrowsers(req);
}
```

            <p>Error loading package location!</p>
          }
        >
          {(packageLocation) => (
            <p>
              Your package is at {packageLocation.latitude}{" "}
              lat and {packageLocation.longitude} long.
            </p>
          )}
        </Await>
      </Suspense>
    </main>
  );
}
```

<details>
  <summary>Alternatively, you can use the `useAsyncValue` hook:</summary>

If you're not jazzed about bringing back render props, you can use a hook, but you'll have to break things out into another component:

```tsx lines=[18,26-31]
import type { UseDataFunctionReturn } from "@remix-run/react";

export default function PackageRoute() {
  const data = useLoaderData<typeof loader>();

  return (
    <main>
      <h1>Let's locate your package</h1>
      <Suspense
        fallback={<p>Loading package location...</p>}
      >
        <Await
          resolve={data.packageLocation}
          errorElement={
            <p>Error loading package location!</p>
          }
        >
          <PackageLocation />
        </Await>
      </Suspense>
    </main>
  );
}

function PackageLocation() {
  const packageLocation =
    useAsyncValue<

>       UseDataFunctionReturn<

Reviewer 5: We renamed this type.

        typeof loader
      >["packageLocation"]
    >();

  return (
    <p>
      Your package is at {packageLocation.latitude} lat and{" "}
      {packageLocation.longitude} long.
    </p>
  );
}
```

</details>

## Evaluating the solution

So rather than waiting for the whole `document -> JavaScript -> hydrate -> request` cycle, with streaming we start the request for the slow data as soon as the document request comes in. This can significantly speed up the user experience.

> Additionally, the API that Remix exposes for this is extremely ergonomic. You can literally switch between whether something is going to be deferred or not based on whether you include the `await` keyword:

Reviewer 5: ```suggestion
Remix treats any `Promise` values as deferred data. These will be streamed to the client as each `Promise` resolves. If you'd like to prevent streaming for critical data, just use `await` and it will be included in the initial presentation of the document to the user.
```

```tsx

> return defer({
>   // not deferred:
>   packageLocation: await packageLocationPromise,
>   // deferred:
>   packageLocation: packageLocationPromise,

Reviewer 5: ```suggestion
return defer({
  // critical data (not deferred):
  packageLocation: await packageLocationPromise,
  // non-critical data (deferred):
  packageLocation: packageLocationPromise,
```

});
```

Because of this, you can A/B test deferring, or even determine whether to defer based on the user or data being requested:

```tsx
export async function loader({
  request,
  params,
}: LoaderArgs) {
  const packageLocationPromise = getPackageLocation(
    params.packageId
  );
  const shouldDefer = await shouldDeferPackageLocation(
    request,
    params.packageId
  );

  return defer({
    packageLocation: shouldDefer
      ? packageLocationPromise
      : await packageLocationPromise,
  });
}
```

That `shouldDeferPackageLocation` could be implemented to check the user making the request, whether the package location data is in a cache, the status of an A/B test, or whatever else you want. This is pretty sweet 🍭

Also, because this happens at request time (even on client transitions), makes use of the URL via nested routing (rather than requiring you to render before you know what data to fetch), and it's all just regular HTTP, we can prefetch and cache the response! Meaning client-side transitions can be _much_ faster (in fact, there are plenty of situations when the user may never be presented with the fallback at all).

> Another powerful thing that's not immediately recognizable is if your server can finish loading deferred data before the client can load the javascript and hydrate, the server will stream down the HTML and "pop" it into place before react is even on the page, increasing performance for those on slow networks.

Reviewer 5: ```suggestion
Another thing that's not immediately recognizable is if your server can finish loading deferred data before the client can load the javascript and hydrate, the server will stream down the HTML and add it to the document *before React hydrates*, thereby increasing performance for those on slow networks. This works even if you never add `<Scripts />` to the page thanks to React 18's support for out-of-order streaming.
```
